src/lib/coda_base/pending_coinbase.ml

```diff
@@ -0,0 +1,346 @@
+    |> of_hash
+
+  let empty_hash =
+    let open Tick.Pedersen in
+    digest_fold
+      (State.create params ~get_chunk_table)
+      (Fold.string_triples "nothing up my sleeve")
+    |> of_hash
+
+  let of_digest = Fn.compose Fn.id of_hash
+
+  type path = Pedersen.Digest.t list
+
+  type _ Request.t +=
+    | Stack_path : Index.t -> path Request.t
+    | Get_coinbase_stack : Index.t -> (Stack.t * path) Request.t
+    | Set_coinbase_stack : Index.t * Stack.t -> unit Request.t
+    | Find_index_of_newest_stack : Index.t Request.t
+    | Find_index_of_oldest_stack : Index.t Request.t
+
+  let reraise_merkle_requests (With {request; respond}) =
+    match request with
+    | Merkle_tree.Get_path addr -> respond (Delegate (Stack_path addr))
+    | Merkle_tree.Set (addr, stack) ->
+        respond (Delegate (Set_coinbase_stack (addr, stack)))
+    | Merkle_tree.Get_element addr ->
+        respond (Delegate (Get_coinbase_stack addr))
+    | _ -> unhandled
+
+  let get t addr =
+    handle
+      (Merkle_tree.get_req ~depth (var_to_hash_packed t) addr)
+      reraise_merkle_requests
+
+  (*
+   [modify_stack t pk ~filter ~f] implements the following spec:
+
+   - finds a coinbase stack [stack] in [t] at path [addr] where [filter stack] holds.
+     note that the stack is not guaranteed to be in the tree in which case it must
+     just have the one coinbase.
+   - returns a root [t'] of a tree of depth [depth]
+   which is [t] but with the stack [f stack] at path [addr].
+*)
+  let%snarkydef modify_stack' t ~(filter : Stack.var -> ('a, _) Checked.t) ~f =
+    let%bind addr =
+      request_witness Index.Unpacked.typ
+        As_prover.(map (return ()) ~f:(fun _ -> Find_index_of_newest_stack))
+    in
+    handle
+      (Merkle_tree.modify_req ~depth (var_to_hash_packed t) addr
+         ~f:(fun stack ->
+           let%bind x = filter stack in
+           f x stack ))
+      reraise_merkle_requests
+    >>| var_of_hash_packed
+
+  (*
+   [edit_stack t pk ~f] implements the following spec:
+
+   - finds a coinbase stack [stack] in [t] at path [addr] OR it doesn't and is a stack with one coinbase
+   - returns a root [t'] of a tree of depth [depth]
+   which is [t] but with the stack [f stack] at path [addr].
+*)
+  let update_stack t ~is_new_stack ~f =
+    modify_stack' t
+      ~filter:(fun stack ->
+        let%bind empty_stack =
+          Stack.Checked.equal stack Stack.(var_of_t empty)
+        in
+        let%bind new_stack = Boolean.(empty_stack && is_new_stack) in
+        let%bind yes = Boolean.(new_stack || not empty_stack) in
+        let%bind () = Boolean.Assert.is_true yes in
+        return yes )
+      ~f:(fun is_empty_or_writeable x -> f ~is_empty_or_writeable x)
+
+  let%snarkydef delete_stack t ~f =
+    let filter stack =
+      let%bind empty_stack =
+        Stack.Checked.equal stack Stack.(var_of_t empty)
+      in
+      let%bind () = Boolean.(Assert.is_true (not empty_stack)) in
```

Comment:
same as above

Reply:
I added an assertion here to know early on if there would be an error. We should not be trying to delete an empty stack.